compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/BIRPackageSymbolEnter.java

```diff
@@ -0,0 +1,1394 @@
+            throw new BLangCompilerException(e.getMessage(), e);
+            //            return null;
+        } catch (Throwable e) {
+            // TODO format error
+            throw new BLangCompilerException(e.getMessage(), e);
+            //            return null;
+        }
+    }
+
+    private BPackageSymbol definePackage(DataInputStream dataInStream) throws IOException {
+        this.typeReader = new BIRTypeReader(dataInStream); //TODO find a better approach than this
+        byte[] magic = new byte[4];
+        dataInStream.read(magic, 0, 4);
+        if (!Arrays.equals(magic, BIRPackageFile.BIR_MAGIC)) {
+            // TODO dlog.error() with package name
+            throw new BLangCompilerException("invalid magic number " + Arrays.toString(magic));
+        }
+
+        int version = dataInStream.readInt();
+        if (version != BIRPackageFile.BIR_VERSION) {
+            // TODO dlog.error() with package name
+            throw new BLangCompilerException("unsupported program file version " + version);
+        }
+
+        // Read constant pool entries of the package info.
+        this.env.constantPool = readConstantPool(dataInStream);
+
+        int pkgCPIndex = dataInStream.readInt();
+        return definePackage(dataInStream, pkgCPIndex);
+    }
+
+    private BPackageSymbol definePackage(DataInputStream dataInStream, int pkgCpIndex) throws IOException {
+
+        PackageCPEntry pkgCpEntry = (PackageCPEntry) this.env.constantPool[pkgCpIndex];
+
+        String orgName = ((StringCPEntry) this.env.constantPool[pkgCpEntry.orgNameCPIndex]).value;
+        String pkgName = ((StringCPEntry) this.env.constantPool[pkgCpEntry.pkgNameCPIndex]).value;
+        String pkgVersion = ((StringCPEntry) this.env.constantPool[pkgCpEntry.versionCPIndex]).value;
+
+        PackageID pkgId = createPackageID(orgName, pkgName, pkgVersion);
+        this.env.pkgSymbol = Symbols.createPackageSymbol(pkgId, this.symTable);
+
+        // TODO Validate this pkdID with the requestedPackageID available in the env.
+
+        // Define import packages.
+        defineSymbols(dataInStream, rethrow(this::defineImportPackage));
+
+        // Define typeDescRef definitions.
+        this.structureTypes = new ArrayList<>();
+        defineSymbols(dataInStream, rethrow(this::defineTypeDef));
+
+        // Define package level variables.
+        defineSymbols(dataInStream, rethrow(this::definePackageLevelVariables));
+
+        defineAttachedFunctions(dataInStream);
+
+
+        // Define annotations.
+//        defineSymbols(dataInStream, rethrow(this::defineAnnotations));
+
+        // Define services.
+//        defineSymbols(dataInStream, rethrow(this::defineService));
+
+        // Resolve unresolved types.
+//        resolveTypes();
+
+        // Read resource info entries.
+//        defineSymbols(dataInStream, rethrow(this::defineResource));
+
+        // Define constants.
+//        defineSymbols(dataInStream, rethrow(this::defineConstants));
```

Comment:
How do we know the number of bytes to be ignored in these commented out cases?

Reply:
those are not serialized yet, only what is serialized is getting deserialized now, I just kept these so that I know these are needed to be symbol entered, anyway there are some code cleanups to be done, will do them in a separate PR